lib/cli/src/Cardano/CLI.hs

```diff
@@ -298,6 +322,133 @@ runCli = join . customExecParser preferences
+-- another type, to include Shelley keys.
+parseKeyType :: Parser (IO ByronWalletStyle)
+parseKeyType = toIO . fromText <$> strOption
+    ( long "type"
+    <> metavar "KEYTYPE"
+    <> helpDoc (Just (vsep typeOptions))
+    )
+  where
+    toIO :: Either TextDecodingError ByronWalletStyle -> IO ByronWalletStyle
+    toIO (Right x) =
+        return x
+    toIO (Left (TextDecodingError e)) =
+        hPutStrLn stderr e >> exitFailure
+
+    typeOptions = string <$>
+        [ "Any of the following:"
+        , "  random (" ++ allowedWords Random ++ ")"
+        , "  icarus (" ++ allowedWords Icarus ++ ")"
+        , "  trezor (" ++ allowedWords Trezor ++ ")"
+        , "  ledger (" ++ allowedWords Ledger ++ ")"
+        ]
+
+    allowedWords = (++ " words")
+        . formatEnglishEnumeration
+        . map show
+        . allowedWordLengths
+        . byronCliKeyScheme
+       where
+          -- >>> formatEnglishEnumeration ["a", "b", "c"]
+          -- "a, b or c"
+          --
+          -- >>> formatEnglishEnumeration ["a", "b"]
+          -- "a or b"
+          --
+          -- >>> formatEnglishEnumeration ["a"]
+          -- "a"
+         formatEnglishEnumeration = formatEnglishEnumerationRev . reverse
+         formatEnglishEnumerationRev [ult, penult]
+            = penult ++ " or " ++ ult
+         formatEnglishEnumerationRev (ult:penult:revBeginning)
+            = intercalate ", " (reverse revBeginning)
+                ++ ", "
+                ++ penult
+                ++ " or "
+                ++ ult
+         formatEnglishEnumerationRev xs = intercalate ", " (reverse xs)
+
+parseSeedWords :: Parser [Text]
+parseSeedWords = some (argument str (metavar "MNEMONIC_WORDS..."))
+
+-- | Lay thine eyes upon the type parameters and see that there are none.
+data CLIKeyScheme = CLIKeyScheme
+    { wordsToSeed :: [Text] -> Either String (Passphrase "seed")
+    , allowedWordLengths :: [Int]
+    , seedToRootKey :: Passphrase "seed" -> XPrv
+    }
+
+byronCliKeyScheme :: ByronWalletStyle -> CLIKeyScheme
+byronCliKeyScheme = \case
+    Random -> mkScheme (Proxy @'Random)
+            (Byron.getKey . flip Byron.generateKeyFromSeed pass)
+    Icarus ->
+        mkScheme (Proxy @'Icarus)
+            icarusStyleKeyToSeed
+    Trezor ->
+        mkScheme (Proxy @'Trezor)
+            icarusStyleKeyToSeed
+    Ledger ->
+        mkScheme (Proxy @'Ledger)
+            icarusStyleKeyToSeed
+  where
+    mkScheme
+        :: forall (s :: ByronWalletStyle).
+            ( FromMnemonic (AllowedMnemonics s) "seed"
+            , ManyNatVal (AllowedMnemonics s)
+            )
+        => Proxy s
+        -> (Passphrase "seed" -> XPrv)
+        -> CLIKeyScheme
+
+    -- Construct a KeyScheme using the type families of the promoted type
+    -- argument.
+    mkScheme _ = CLIKeyScheme
+            (left getFromMnemonicError . fromMnemonic @(AllowedMnemonics s) @"seed")
+            (map fromIntegral (manyNatVal $ Proxy @(AllowedMnemonics s)))
```

Comment:
Somehow this indent (and others like it) ended up being 8 spaces instead of 4 spaces.

Perhaps reduce this back to 4 spaces?